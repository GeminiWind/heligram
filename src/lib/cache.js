import { promisify } from 'util';
import redis from 'redis';

const client = redis.createClient('redis://cache');

const get = promisify(client.get).bind(client);
const set = promisify(client.set).bind(client);
const hset = promisify(client.hset).bind(client);
const hget = promisify(client.hget).bind(client);
const del = promisify(client.del).bind(client);
const hdel = promisify(client.hdel).bind(client);

export default {
  get,
  set,
  hset,
  hget,
  del,
  hdel,
};
