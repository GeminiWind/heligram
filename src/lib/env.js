import { config } from 'dotenv';

export default function loadEnv() {
  const envFilePath = process.env.NODE_ENV ? `env.${process.env.NODE_ENV}` : '.env.development';

  config({ path: envFilePath });
}
