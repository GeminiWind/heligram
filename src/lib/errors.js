import JsonApiError from './jsonApiError';

class InternalError extends JsonApiError {
  constructor(message) {
    let internalError = {
      title: 'InternalError',
      code: 'InternalError',
      status: 500,
      detail: 'InternalError',
    };
    if (typeof message === 'string') {
      internalError = {
        ...internalError,
        detail: message,
      };
    }
    if (typeof message === 'object') {
      internalError = {
        ...internalError,
        ...message,
      };
    }
    super(internalError);
  }
}

class NotFoundError extends JsonApiError {
  constructor(message) {
    let notFoundError = {
      title: 'NotFoundError',
      code: 'NotFoundError',
      status: 404,
      detail: 'NotFoundError',
    };
    if (typeof message === 'string') {
      notFoundError = {
        ...notFoundError,
        detail: message,
      };
    }
    if (typeof message === 'object') {
      notFoundError = {
        ...notFoundError,
        ...message,
      };
    }
    super(notFoundError);
  }
}

class BadRequestError extends JsonApiError {
  constructor(message) {
    let badRequestError = {
      title: 'BadRequestError',
      code: 'BadRequestError',
      status: 400,
      detail: 'BadRequestError',
    };
    if (typeof message === 'string') {
      badRequestError = {
        ...badRequestError,
        detail: message,
      };
    }
    if (typeof message === 'object') {
      badRequestError = {
        ...badRequestError,
        ...message,
      };
    }
    super(badRequestError);
  }
}

class MalformedError extends JsonApiError {
  constructor() {
    const malformedError = {
      title: 'MalformedError',
      code: 'MalformedError',
      status: 400,
      detail: 'Request body is malformed.',
    };
    super(malformedError);
  }
}

export {
  InternalError,
  NotFoundError,
  BadRequestError,
  MalformedError,
};
