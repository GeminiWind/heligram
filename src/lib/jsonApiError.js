class JsonApiError extends Error {
  constructor(error) {
    const errorMessage = `${error.title} ${error.detail}`;
    super(errorMessage);

    this.title = error.title;
    this.detail = error.detail;
    this.status = error.status;
    this.code = error.code;

    if (error.source) {
      this.source = error.source;
    }
    if (error.meta) {
      this.meta = error.meta;
    }

    // Ensure the name of this error is the same as the class name
    this.name = this.constructor.name;
    // This clips the constructor invocation from the stack trace.
    // It's not absolutely essential, but it does make the stack trace a little nicer.
    //  @see Node.js reference (bottom)
    Error.captureStackTrace(this, this.constructor);
  }

  toJSON() {
    return {
      status: this.status,
      code: this.code,
      title: this.title,
      detail: this.detail,
      ...(this.source ? { source: this.source } : {}),
      ...(this.meta ? { source: this.meta } : {}),
    };
  }
}

export default JsonApiError;
