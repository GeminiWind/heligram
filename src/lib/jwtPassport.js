import { Strategy as JwtStrategy, ExtractJwt } from 'passport-jwt';
import storageLibrary from './storageLibrary';

export default function jwtPassport(passport) {
  const options = {
    jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
    secretOrKey: process.env.APP_KEY,
  };

  passport.use(new JwtStrategy(options, (jwtPayload, done) => {
    const {
      email,
      scopes,
    } = jwtPayload;

    storageLibrary.findOne({
      Path: `user/${email}`,
    }).then((record) => {
      if (record) {
        const {
          Content: {
            password,
            ...userAttributesWithoutPassword
          },
        } = record;
        done(null, {
          ...userAttributesWithoutPassword,
          scopes,
        });
      } else done(null, false);
    }).catch(error => done(error, false));
  }));
}
