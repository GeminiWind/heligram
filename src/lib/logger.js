import moment from 'moment';
import path from 'path';
import { format, transports, createLogger } from 'winston';

const logger = createLogger({
  format: format.combine(
    format.splat(),
    format.timestamp({
      format: 'YYYY-MM-DD HH:mm:ss',
    }),
    format.printf(info => `${info.timestamp} ${info.level}: ${info.message}`),
  ),
  transports: [
    new transports.Console({
      format: format.combine(
        format.colorize(),
        format.timestamp({
          format: 'YYYY-MM-DD HH:mm:ss',
        }),
        format.printf(info => `${info.timestamp} ${info.level}: ${info.message}`),
      ),
    }),
    new transports.File({
      filename: path.resolve(__dirname, '../..', 'log', `${moment().format('YYYY-MM-DD')}.log`),
    }),
  ],
});

export default logger;
