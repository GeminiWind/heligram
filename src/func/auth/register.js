import bcrypt from 'bcrypt';
import { BadRequestError, InternalError } from '../../lib/errors';

function validateRequest(req) {
  const { schemaValidator, instrumentation } = req;
  const isValid = schemaValidator.validate('https://heligram.com/create-user+v1.json', req.body);

  if (!isValid) {
    instrumentation.error('Error in validate request %s', JSON.stringify(schemaValidator.errors, null, 2));

    throw new BadRequestError({
      source: schemaValidator.errors,
    });
  }

  return req;
}

async function isUserEmailExist(req) {
  const {
    instrumentation,
    storageLibrary,
    body: {
      data: {
        attributes: {
          email,
        },
      },
    },
  } = req;

  const user = await storageLibrary.findOne({
    Path: `user/${email}`,
  });

  if (user) {
    instrumentation.error(`User with ${email} already exist.`);

    throw new BadRequestError(`User with ${email} already exist.`);
  }

  return req;
}

async function createUser(req) {
  const {
    instrumentation,
    storageLibrary,
    body: {
      data,
    },
  } = req;

  const saltRounds = 10;
  const salt = bcrypt.genSaltSync(saltRounds);
  const hashedPassword = bcrypt.hashSync(data.password, salt);

  const record = {
    Path: `user/${data.attributes.email}`,
    Content: {
      email: data.attributes.email,
      password: hashedPassword,
      scopes: 'user:profile create:chat read:chat',
    },
    Type: 'users',
  };

  try {
    await storageLibrary.create(record);
  } catch (error) {
    instrumentation.error('Error in creating user.', error);

    throw new InternalError('Error in creating user.');
  }

  return req;
}

function returnResponse(req) {
  return {
    statusCode: 201,
    body: {
      data: {
        type: 'users',
        attributes: {
          email: req.body.data.attributes.email,
        },
      },
    },
  };
}

export default req => Promise.resolve(req)
  .then(validateRequest)
  .then(isUserEmailExist)
  .then(createUser)
  .then(returnResponse);
