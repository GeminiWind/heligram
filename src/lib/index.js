import logger from './logger';
import storageLibrary from './storageLibrary';

export { logger, storageLibrary };
