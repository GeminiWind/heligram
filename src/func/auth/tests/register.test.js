import register, {
  validateRequest,
  isUserEmailExist,
  createUser,
  returnResponse,
} from '../register';
import { schemaValidator } from '../../../lib';
import { BadRequestError, InternalError } from '../../../lib/errors';

describe('Register Controller', () => {
  it('return original request if request is valid with corresponding schema', () => {
    const request = {
      body: {
        data: {
          type: 'users',
          attributes: {
            email: 'validEmail@example.com',
            password: 'validPassword',
          },
        },
      },
      instrumentation: {
        error: jest.fn(),
      },
      schemaValidator,
    };

    let error;

    try {
      validateRequest(request);
    } catch (e) {
      error = e;
    }

    expect(error).toBeFalsy();
  });

  it('throw \'BadRequestError\'  in case that request does not matches with schema', () => {
    const request = {
      body: {
        data: {
          type: 'invalid-type',
          attributes: {
            email: 'validEmail@example.com',
            password: 'validPassword',
          },
        },
      },
      instrumentation: {
        error: jest.fn(),
      },
      schemaValidator,
    };

    let error;

    try {
      validateRequest(request);
    } catch (e) {
      error = e;
    }

    expect(error).toBeInstanceOf(BadRequestError);
    expect(error).toMatchSnapshot();
  });

  it('return origin request if intended user email does not exist', async () => {
    const request = {
      body: {
        data: {
          type: 'users',
          attributes: {
            email: 'validEmail@example.com',
            password: 'validPassword',
          },
        },
      },
      instrumentation: {
        error: jest.fn(),
      },
      schemaValidator,
      storageLibrary: {
        findOne: jest.fn().mockImplementation(() => ({
          cache: jest.fn().mockImplementation(() => Promise.resolve(false)),
        })),
      },
    };

    let error;

    try {
      await isUserEmailExist(request);
    } catch (e) {
      error = e;
    }

    expect(error).toBeUndefined();
  });

  it('throw \'BadRequestError\' if intended user email does exist', async () => {
    const request = {
      body: {
        data: {
          type: 'invalid-type',
          attributes: {
            email: 'emailExist@example.com',
            password: 'validPassword',
          },
        },
      },
      instrumentation: {
        error: jest.fn(),
      },
      schemaValidator,
      storageLibrary: {
        findOne: jest.fn().mockImplementation(() => ({
          cache: jest.fn().mockImplementation(() => Promise.resolve({
            Type: 'users',
            Content: {
              email: 'emailExist@example.com',
              password: 'fake-hashed-password',
            },
            Attributes: {
              createdAt: '2019-12-12 08:00:00+07:00',
              updatedAt: '2019-12-12 08:00:00+07:00',
            },
          })),
        })),
      },
    };

    let error;

    try {
      await isUserEmailExist(request);
    } catch (e) {
      error = e;
    }

    expect(error).toBeInstanceOf(BadRequestError);
    expect(error).toMatchSnapshot();
  });

  it('can create new user', async () => {
    const request = {
      body: {
        data: {
          type: 'users',
          attributes: {
            email: 'emailExist@example.com',
            password: 'validPassword',
          },
        },
      },
      instrumentation: {
        error: jest.fn(),
      },
      storageLibrary: {
        create: jest.fn().mockImplementation(() => Promise.resolve(true)),
      },
      cache: {
        del: jest.fn().mockImplementation(() => Promise.resolve(true)),
      },
    };

    let error;

    try {
      await createUser(request);
    } catch (e) {
      error = e;
    }

    expect(error).toBeUndefined();
    expect(request.storageLibrary.create.mock.calls.length).toBe(1);
    expect(request.cache.del.mock.calls.length).toBe(1);
  });

  it('throw \'InternalError\' if it encounters any error when saving record in database ', async () => {
    const request = {
      body: {
        data: {
          type: 'users',
          attributes: {
            email: 'emailExist@example.com',
            password: 'validPassword',
          },
        },
      },
      instrumentation: {
        error: jest.fn(),
      },
      storageLibrary: {
        create: jest.fn().mockImplementation(() => Promise.reject(new Error('Runtime error'))),
      },
      cache: {
        del: jest.fn().mockImplementation(() => Promise.resolve(true)),
      },
    };

    let error;

    try {
      await createUser(request);
    } catch (e) {
      error = e;
    }

    expect(error).toBeInstanceOf(InternalError);
    expect(request.storageLibrary.create.mock.calls.length).toBe(1);
    expect(request.instrumentation.error.mock.calls.length).toBe(1);
    expect(request.cache.del.mock.calls.length).toBe(0);
  });

  it('can return response correctly', () => {
    const request = {
      body: {
        data: {
          type: 'invalid-type',
          attributes: {
            email: 'emailExist@example.com',
            password: 'validPassword',
          },
        },
      },
    };

    const response = returnResponse(request);

    expect(response).toHaveProperty('statusCode', 201);
    expect(response).toHaveProperty('body');
    expect(response).toMatchSnapshot();
  });

  it('create new user successfully as whole flow', async () => {
    const request = {
      body: {
        data: {
          type: 'users',
          attributes: {
            email: 'emailExist@example.com',
            password: 'validPassword',
          },
        },
      },
      instrumentation: {
        error: jest.fn(),
      },
      schemaValidator,
      storageLibrary: {
        findOne: jest.fn().mockImplementation(() => ({
          cache: jest.fn().mockImplementation(() => Promise.resolve(false)),
        })),
        create: jest.fn().mockImplementation(() => Promise.resolve(true)),
      },
      cache: {
        del: jest.fn().mockImplementation(() => Promise.resolve(true)),
      },
    };

    const response = await register(request);

    expect(response).toHaveProperty('statusCode', 201);
    expect(response).toHaveProperty('body');
    expect(response).toMatchSnapshot();
  });
});
