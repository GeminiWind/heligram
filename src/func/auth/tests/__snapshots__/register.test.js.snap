// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Controller can return response correctly 1`] = `
Object {
  "body": Object {
    "data": Object {
      "attributes": Object {
        "email": "emailExist@example.com",
      },
      "type": "users",
    },
  },
  "statusCode": 201,
}
`;

exports[`Register Controller create new user successfully as whole flow 1`] = `
Object {
  "body": Object {
    "data": Object {
      "attributes": Object {
        "email": "emailExist@example.com",
      },
      "type": "users",
    },
  },
  "statusCode": 201,
}
`;

exports[`Register Controller throw 'BadRequestError'  in case that request does not matches with schema 1`] = `[BadRequestError: BadRequestError BadRequestError]`;

exports[`Register Controller throw 'BadRequestError' if intended user email does exist 1`] = `[BadRequestError: BadRequestError User with emailExist@example.com already exist.]`;
